### ISSUE TITLE
IncrementalBuild object has already been disposed!

### ISSUE THREAD

**Issue by vasily-kirichenko** [Microsoft/visualfsharp#2472]
![image](https://cloud.githubusercontent.com/assets/873919/23231056/6db0c254-f957-11e6-87f2-5fb8a89b6f14.png)

```
---------------------------
Assertion Failed: Abort=Quit, Retry=Debug, Ignore=Continue
---------------------------
IncrementalBuild object has already been disposed!



   at Microsoft.FSharp.Compiler.IncrementalBuilder.assertNotDisposed()

   at Microsoft.FSharp.Compiler.IncrementalBuilder.StampFileNameTask[a,b](TimeStampCache cache, a _ctok, range _m, String filename, b _isLastCompiland)

   at <StartupCode$FSharp-LanguageService-Compiler>.$IncrementalBuild.-ctor@1579-78.Invoke(TimeStampCache cache, CompilationThreadToken _ctok, Tuple`3 tupledArg)

   at Microsoft.FSharp.Compiler.IncrementalBuild.CheckStamp@711.Invoke(DateTime acc, Int32 slot)

   at Microsoft.FSharp.Collections.ListModule.loop@220-16.Invoke(TState s, FSharpList`1 xs)

   at Microsoft.FSharp.Compiler.IncrementalBuilder.GetLogicalTimeStampForProject(TimeStampCache cache, CompilationThreadToken ctok)

   at Microsoft.FSharp.Compiler.SourceCodeServices.BackgroundCompiler.TryGetLogicalTimeStampForProject(TimeStampCache cache, CompilationThreadToken ctok, FSharpProjectOptions options)

   at <StartupCode$FSharp-LanguageService-Compiler>.$Service.projectReferences@2244-1.Microsoft-FSharp-Compiler-CompileOps-IProjectReference-TryGetLogicalTimeStamp(TimeStampCache , CompilationThreadToken )

   at Microsoft.FSharp.Compiler.CompileOps.TimeStampCache.GetProjectReferenceTimeStamp(IProjectReference pr, CompilationThreadToken ctok)

   at <StartupCode$FSharp-LanguageService-Compiler>.$IncrementalBuild.nonFrameworkAssemblyInputs@1897-2.Invoke(TimeStampCache cache, CompilationThreadToken ctok)

   at Microsoft.FSharp.Compiler.IncrementalBuild.checkStamp@626.Invoke(T acc, Int32 slot)

   at Microsoft.FSharp.Collections.ListModule.loop@220-16.Invoke(TState s, FSharpList`1 xs)

   at Microsoft.FSharp.Compiler.IncrementalBuild.visitVector@541-1.Invoke(FSharpOption`1 optSlot, VectorBuildRule ve, T acc)

   at Microsoft.FSharp.Compiler.IncrementalBuild.visitVector@541-1.Invoke(FSharpOption`1 optSlot, VectorBuildRule ve, T acc)

   at Microsoft.FSharp.Compiler.IncrementalBuild.Step@805.Invoke(Unit unitVar)

   at Microsoft.FSharp.Compiler.AbstractIL.Internal.Library.CancellableModule.delay@576.Invoke(CancellationToken ct)

   at Microsoft.FSharp.Compiler.AbstractIL.Internal.Library.CancellableModule.bind@543.Invoke(CancellationToken ct)

   at Microsoft.FSharp.Compiler.AbstractIL.Internal.Library.CancellableModule.runWithoutCancellation[a](Cancellable`1 comp)

   at <StartupCode$FSharp-LanguageService-Compiler>.$Service.CheckProjectInBackground@2762.Invoke(CompilationThreadToken ctok)

   at <StartupCode$FSharp-LanguageService-Compiler>.$Reactor.loop@67-90.Invoke(FSharpOption`1 _arg2)

   at Microsoft.FSharp.Control.AsyncBuilderImpl.cont@825.Invoke(a a)

   at <StartupCode$FSharp-Core>.$Control.loop@428-38.Invoke(FSharpFunc`2 action)

   at Microsoft.FSharp.Control.Trampoline.ExecuteAction(FSharpFunc`2 firstAction)

   at Microsoft.FSharp.Control.TrampolineHolder.Protect(FSharpFunc`2 firstAction)

   at <StartupCode$FSharp-Core>.$Control.AwaitWaitHandle@1760-3.Invoke(Object state, Boolean timedOut)

   at System.Threading._ThreadPoolWaitOrTimerCallback.WaitOrTimerCallback_Context(Object state, Boolean timedOut)

   at System.Threading._ThreadPoolWaitOrTimerCallback.WaitOrTimerCallback_Context_t(Object state)

   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)

   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)

   at System.Threading._ThreadPoolWaitOrTimerCallback.PerformWaitOrTimerCallback(Object state, Boolean timedOut)


---------------------------
Abort   Retry   Ignore   
---------------------------
```

**Comment by vasily-kirichenko** [comment]
It pops up endlessly, VS is unusable.

**Comment by dsyme** [comment]
I'll review the last PR for Cancellable.  Do you have any details on the solution you're loading?

**Comment by dsyme** [comment]
@vasily-kirichenko I've reviewed the changes and I don't see the mistake yet https://github.com/Microsoft/visualfsharp/pull/2460/files.  

[This call](https://github.com/Microsoft/visualfsharp/pull/2460/files#diff-d3c6e872b60cb8a91665cdd9d5e5ca27R2673) should have incremented the usage count.  

Is it a reliable repro?  Was the above the callstack of the first assert? 

thanks

**Comment by dsyme** [comment]
Ugh I'm stumped.

Some things we could do

* Put an assert in ``getOrCreateBuilder`` that checks that if we get it from the ``incrementalBuildersCache`` then the thing is not already disposed (there should never be ``disposed`` entries accessible from that store).  But that won't do anything more than shift the assert one step earlier (I presume

* Put some tracing data in ``IncrementaBuilder`` that records the history of callstacks that have done increment/decrement operations.  In debug mode we could store the last 50-100 callstacks in each builder.  

If this is happening all the time (reliable repro) then surely there was a basic mistake in the last PR of mine.  If it's happening only very occasionally then it was probably something much more subtle, perhaps a pre-existing problem, and we may need to use the second technique above to find the problem.

**Comment by KevinRansom** [comment]
@dsyme it is not hard to repro.  But I think the exception happened before on debug builds.  I will go back a day or two, and try it out.

**Comment by dsyme** [comment]
@KevinRansom Thanks.  We need to get to the bottom of it as with type providers it's quite a severe problem (otherwise the problem is pretty harmless).  Perhaps we should start running some VS tests in the debug CI build.

I think @vasily-kirichenko may have been regularly using debug builds

**Comment by KevinRansom** [comment]
@dsyme looking at it :-)

**Comment by KevinRansom** [comment]
I was correct, Huh!!! how often is that the case ?

 :-)
Here without the cancellable callstack.

![image](https://cloud.githubusercontent.com/assets/5175830/23235934/31e6d0f6-f90c-11e6-8929-c622cba43e00.png)

**Comment by KevinRansom** [comment]
And this is the stacktrace with the cancellable on the callstack:  Note the ErrorLogger+InternalError for both.

![image](https://cloud.githubusercontent.com/assets/5175830/23237040/7cc5e626-f910-11e6-8d5d-8795d9559df5.png)

**Comment by dsyme** [comment]
@KevinRansom That's a different throw/assert - it's in TcImports (one of only a handful of disposable objects in the compiler) rather than IncrementalBuilder.  But it may be related still. What's the repro to get those stacktraces? thanks

**Comment by KevinRansom** [comment]
easy install the debug .vsix.  Open visualfsharp sln and open a .fs file, click about a bit and it will go bang.

**Comment by dsyme** [comment]
> easy install the debug .vsix. Open visualfsharp sln and open a .fs file, click about a bit and it will go bang.

@KevinRansom Is this problem gone after #2479 ? Thanks.  

@vasily-kirichenko Please let me know if the "IncrementalBuild object has already been disposed!" problem recurs (which I think is different to the problem reported by @KevinRansom though there may be similar underlying causes)

**Comment by vasily-kirichenko** [comment]
@dsyme I've not seen "already been disposed" so far, but a new one popped up - "FSharp.Core was not delayed" (not precisely this text).

**Comment by dsyme** [comment]
> @dsyme I've not seen "already been disposed" so far, but a new one popped up - "FSharp.Core was not delayed" (not precisely this text).

If you could capture any possible information when that happens that would be great

**Comment by cartermp** [comment]
@vasily-kirichenko Are you still seeing this?

**Comment by vasily-kirichenko** [comment]
No.

**Comment by cartermp** [comment]
😄
